Unable to build repo on latest VS dogfood

I've just sync'd to the latest in main and cleaned my repo (`git clean -xdf`).  I have VS Version 16.10.0 Preview 2.0 [31115.18.main] installed.

When I try to build (`.\build.cmd clr+libs -rc release`), I get ~400 warnings of the form:
```
  cl : command line warning D9002: ignoring unknown option '/external:envEXTERNAL_INCLUDES' [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\jit\static\clrjit_obj.vcxproj]
```
and then an error of the form:
```
LINK : fatal error LNK1268: inconsistent option 'ILK:coreclr.dir\Release\coreclr.ilk' specified with /USEPROFILE but not with /GENPROFILE [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\dlls\mscoree\coreclr\coreclr.vcxproj]
```
that kills the build.

@jkotas, I'm on the latest VS build that's offered to me (31118.392.main) and I'm sync'd to main, but I'm now unable to build, even with -ninja.  What version of VS do you have?

I hit linking errors on ildasm:
```
  hotheapwriter.cpp.obj : error LNK2011: precompiled object not linked in; image may not run
    Hint on symbols that are defined and could potentially match:
      __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUmnzpvlyqUdrmwldhOcGEOivovzhvUfgroxlwvUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvPkxsOxccOlyq@4B2008FD98C1DD4
  regmeta_imetadatatables.cpp.obj : error LNK2011: precompiled object not linked in; image may not run
    Hint on symbols that are defined and could potentially match:
      __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUmnzpvlyqUdrmwldhOcGEOivovzhvUfgroxlwvUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvPkxsOxccOlyq@4B2008FD98C1DD4
  strongnameinternal.cpp.obj : error LNK2011: precompiled object not linked in; image may not run
    Hint on symbols that are defined and could potentially match:
      __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUmnzpvlyqUdrmwldhOcGEOivovzhvUfgroxlwvUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvPkxsOxccOlyq@4B2008FD98C1DD4
  mdinternalrw.cpp.obj : error LNK2011: precompiled object not linked in; image may not run
    Hint on symbols that are defined and could potentially match:
      __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUmnzpvlyqUdrmwldhOcGEOivovzhvUfgroxlwvUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvurovhUfgroxlwvhgzgrxmlslhgOwriUxnzpvPkxsOxccOlyq@4B2008FD98C1DD4
  ildasm\exe\ildasm.exe : fatal error LNK1120: 4 unresolved externals
```

Without ninja, I get:
```
LINK : fatal error LNK1268: inconsistent option 'ILK:coreclr.dir\Release\coreclr.ilk' specified with /USEPROFILE but not with /GENPROFILE [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\dlls\mscoree\coreclr\coreclr.vcxproj]
```
Is that the warning-as-error you're referring to?

With ninja, I get the aforementioned `_@@_PchSym_` errors.

This is after doing:
git fetch --all
git reset --hard upstream/main
git clean -xdf

> fatal error LNK1268: inconsistent option 'ILK:coreclr.dir\Release\coreclr.ilk' specified with /USEPROFILE but not with /GENPROFILE

Yes, that's the error that I am referring to. It is actually non-surprissible hard error and I have not found yet what needs to be changed to make it go away.

> With ninja, I get the aforementioned _@@_PchSym_ errors.

I have tried again with ninja and it works fine for me. (I have done `build.cmd -s clr+libs -c Release -ninja`.) What is the exact command that you are using?

I suspect that it may be a build race condition that only repros with ninja. @jkoritzinsky Could you please help to troubleshoot it?

> fatal error LNK1268: inconsistent option 'ILK:coreclr.dir\Release\coreclr.ilk' specified with /USEPROFILE but not with /GENPROFILE

This error can be worked around by passing `/p:NoPgoOptimize=true` to the build, e.g. `build.cmd -s clr+libs -c Release /p:NoPgoOptimize=true`

> This error can be worked around by passing /p:NoPgoOptimize=true to the build, e.g. build.cmd -s clr+libs -c Release /p:NoPgoOptimize=true

No dice.  I did:
```
.\build.cmd -s clr+libs -c Release /p:NoPgoOptimize=true
```
and I got:
```
  Build FAILED.

  hotheapwriter.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
  regmeta_imetadatatables.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
  strongnameinternal.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
  mdinternalrw.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
  D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\Release\ildasm.exe : fatal error LNK1120: 4 unresolved externals [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
  pesectionman.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ilasm\ilasm.vcxproj]
  D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ilasm\Release\ilasm.exe : fatal error LNK1120: 1 unresolved externals [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ilasm\ilasm.vcxproj]
      0 Warning(s)
      7 Error(s)
```
with earlier details like:
```
  hotheapwriter.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
      Hint on symbols that are defined and could potentially match:
        __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUlyqUxlivxoiUdrmwldhOcGEOivovzhvUfgroxlwvUfgroxlwvhgzgrxmlslhgOwriUivovzhvUxnzpvPkxsOlyq@4B2008FD98C1DD4
  regmeta_imetadatatables.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
      Hint on symbols that are defined and could potentially match:
        __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUlyqUxlivxoiUdrmwldhOcGEOivovzhvUfgroxlwvUfgroxlwvhgzgrxmlslhgOwriUivovzhvUxnzpvPkxsOlyq@4B2008FD98C1DD4
  strongnameinternal.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
      Hint on symbols that are defined and could potentially match:
        __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUlyqUxlivxoiUdrmwldhOcGEOivovzhvUfgroxlwvUfgroxlwvhgzgrxmlslhgOwriUivovzhvUxnzpvPkxsOlyq@4B2008FD98C1DD4
  mdinternalrw.obj : error LNK2011: precompiled object not linked in; image may not run [D:\repos\runtime\artifacts\obj\coreclr\windows.x64.Release\ildasm\exe\ildasm.vcxproj]
      Hint on symbols that are defined and could potentially match:
        __@@_PchSym_@00@UivklhUifmgrnvUzigruzxghUlyqUxlivxoiUdrmwldhOcGEOivovzhvUfgroxlwvUfgroxlwvhgzgrxmlslhgOwriUivovzhvUxnzpvPkxsOlyq@4B2008FD98C1DD4
```
No idea how my system differs from yours.

--- a/docs/workflow/requirements/windows-requirements.md
+++ b/docs/workflow/requirements/windows-requirements.md
@@ -42,7 +42,7 @@ The dotnet/runtime repository requires at least Visual Studio 2019 16.6.
 - Add its location (e.g. C:\Program Files (x86)\CMake\bin) to the PATH environment variable.
   The installation script has a check box to do this, but you can do it yourself after the fact following the instructions at [Adding to the Default PATH variable](#adding-to-the-default-path-variable).
 
-The dotnet/runtime repository recommends using CMake 3.16.0 or newer, but works with CMake 3.15.5.
+The dotnet/runtime repository recommends using CMake 3.16.4 or newer, but it may work with CMake 3.15.5.
 
 ## Ninja (optional)
 

--- a/eng/native/set-cmake-path.ps1
+++ b/eng/native/set-cmake-path.ps1
@@ -38,8 +38,8 @@ try {
   if ($version -lt [Version]"3.14.0") {
     throw "This repository requires CMake 3.14 and recommends CMake 3.16. The version of CMake installed is $version. Please install CMake 3.16 or newer from https://cmake.org/download/."
   }
-  elseif ($version -lt [Version]"3.16.0") {
-    [System.Console]::Error.WriteLine("CMake 3.16 or newer is recommended for building this repository. The newest version of CMake installed is $version. Please install CMake 3.16 or newer from https://cmake.org/download/.")
+  elseif ($version -lt [Version]"3.16.4") {
+    [System.Console]::Error.WriteLine("CMake 3.16.4 or newer is recommended for building this repository. The newest version of CMake installed is $version. Please install CMake 3.16.4 or newer from https://cmake.org/download/.")
   }
 
   [System.Console]::WriteLine("set CMakePath=" + $cmakePath)

